console/client/src/layout/IDELayout.tsx

```diff
@@ -1,66 +1,109 @@
-      const index = tabs.findIndex(tab => tab.id === id)
-      setActiveTab(index - 1)
-      searchParams.delete(TabSearchParams.active)
+  const handleCloseTab = (id: string, index: number) => {
+    const nextActiveTab = {
+      id: tabs[index - 1].id,
+      type: tabs[index - 1].type,
     }
+    setSearchParams({
+      ...Object.fromEntries(searchParams),
+      [TabSearchParams.id]: nextActiveTab.id,
+      [TabSearchParams.type]: nextActiveTab.type,
+    })
+    setActiveTab(nextActiveTab)
     setTabs(tabs.filter(tab => tab.id !== id))
-    navigate({...location, search: searchParams.toString()})
   }
 
   const handleChangeTab = (index: number) => {
-    setActiveTab(index)
-    index > 0 && tabs.length > 1
-      ? searchParams.set(TabSearchParams.active, tabs[index].id)
-      : searchParams.delete(TabSearchParams.active)
-    navigate({...location, search: searchParams.toString()})
+    const nextActiveTab = tabs[index]
+    setActiveTab({id: nextActiveTab.id, type: nextActiveTab.type})
+    setSearchParams({
+      ...Object.fromEntries(searchParams),
+      [TabSearchParams.id]: nextActiveTab.id,
+      [TabSearchParams.type]: nextActiveTab.type,
+    })
   }
 
-  // Handle opening the correct tab on load
   React.useEffect(() => {
-    const id = searchParams.get(TabSearchParams.active)
-    if (!id) {
-      setActiveTab(0)
-      return
-    }
-    const index = tabs.findIndex(tab => tab.id === id)
-    if (index >= 0) {
-      setActiveTab(activeTab ?? index)
-      return
-    }
-    const [_, label] = id.split('.')
-    const newTab = {
-      id,
-      label,
-      type: TabType.Verb,
+    const validateTabs = async () => {
+      const modules = await client.getModules({})
+      const msg = invalidTab({id, type})
+      if (msg) {
+        // IDs an invalid tab ID and type fallback to timeline
+        setActiveTab({id: timelineTab.id, type: timelineTab.type})
+        // On intial mount we have no query params set for tabs so we want to skip setting invalidTabMessage
+        if (type === null && id === null) return
+        return setInvalidTabMessage(msg)
+      }
+      const inTabsList = tabs.some(
+        ({id: tabId, type: tabType}) => tabId === id && tabType === type
+      )
+      // Tab is in tab list just set active tab
+      if (inTabsList) return setActiveTab({id, type})
+      // Get module and Verb ids
+      const [moduleId, verbId] = id.split('.')
+      // Check to see if they exist on controller
+      const moduleExist = modules.modules.find(
+        module => module?.name === moduleId
+      )
+      const verbExist = moduleExist?.verbs.some(
+        ({verb}) => verb?.name === verbId
+      )
+      // Set tab if they both exists
+      if (moduleExist && verbExist) {
+        const newTab = {
+          id: moduleId,
+          label: verbId,
+          type: TabType.Verb,
+        }
+        const nextTabs = [...tabs, newTab]
+        return setTabs(nextTabs)
```

Comment:
Should we be setting the active tab before returning here? This seems to leave the UI in an odd state.

<img width="1285" alt="Screenshot 2023-08-28 at 12 55 47 PM" src="https://github.com/TBD54566975/ftl/assets/51647/22f88c6d-f08f-4695-87bc-e6b69de59e34">

Reply:
yep it's missing added it back

Reply:
Hmm, weird. I still don't get the active tab set when I test with the latest code:

Is it because you're setting the tab before adding it to the tab list? Does it work ok for you? Maybe I'm missing something.

<img width="1285" alt="Screenshot 2023-08-28 at 12 55 47 PM" src="https://github.com/TBD54566975/ftl/assets/51647/be2dae07-e7f7-4b29-9dfe-3b30d36eb77f">